tests/integration/pilot/tunneling/tunneling_test.go

```diff
@@ -0,0 +1,416 @@
+	istioResourcesToApply []string
+}
+
+func testCases() []tunnelingTestCase {
+	return []tunnelingTestCase{
+		{
+			"gateway/tcp",
+			[]string{"http", "https"},
+			[]string{
+				"gateway/tcp/virtual-service.yaml",
+				"gateway/tcp/gateway.yaml",
+			},
+		},
+		//{
+		//	"gateway/tls/istio-mutual",
+		//	[]string{"http", "https"},
+		//	[]string{
+		//		"gateway/tls-istio-mutual/mtls.yaml",
+		//		"gateway/tls-istio-mutual/virtual-service.yaml",
+		//		"gateway/tls-istio-mutual/gateway.yaml",
+		//	},
+		//},
+		{
+			"gateway/tls/passthrough",
+			[]string{"https"},
+			[]string{
+				"gateway/tls-passthrough/virtual-service.yaml",
+				"gateway/tls-passthrough/gateway.yaml",
+			},
+		},
+		{
+			"sidecar",
+			[]string{"http", "https"},
+			[]string{
+				"sidecar/virtual-service.yaml",
+			},
+		},
+	}
+}
+
+var i istio.Instance
+
+const (
+	http1 = "HTTP1"
+	http2 = "HTTP2"
+)
+
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite(m).
+		Setup(istio.Setup(&i, enableRegistryOnlyMode)).
+		Run()
+}
+
+func enableRegistryOnlyMode(_ resource.Context, cfg *istio.Config) {
+	cfg.ControlPlaneValues = `
+meshConfig:
+  accessLogFile: /dev/stdout
+  outboundTrafficPolicy:
+    mode: REGISTRY_ONLY`
+}
+
+func TestTunnelingViaHTTP1Proxy(t *testing.T) {
+	framework.
+		NewTest(t).
+		Features("traffic.tunneling.http1_proxy.plain_text").
+		Run(func(ctx framework.TestContext) {
+			runTunnelingTests(t, ctx, http1, false)
+		})
+}
+
+func TestTunnelingViaHTTP1ProxyWithTLS(t *testing.T) {
+	framework.
+		NewTest(t).
+		Features("traffic.tunneling.http1_proxy.tls").
+		Run(func(ctx framework.TestContext) {
+			runTunnelingTests(t, ctx, http1, true)
+		})
+}
+
+func TestTunnelingViaHTTP2Proxy(t *testing.T) {
+	framework.
+		NewTest(t).
+		Features("traffic.tunneling.http2_proxy.plain_text").
+		Run(func(ctx framework.TestContext) {
+			runTunnelingTests(t, ctx, http2, false)
+		})
+}
+
+func TestTunnelingViaHTTP2ProxyWithTLS(t *testing.T) {
+	framework.
+		NewTest(t).
+		Features("traffic.tunneling.http2_proxy.tls").
+		Run(func(ctx framework.TestContext) {
+			runTunnelingTests(t, ctx, http2, true)
+		})
+}
+
+func runTunnelingTests(t *testing.T, ctx framework.TestContext, proxyHTTPVersion string, proxyTLSEnabled bool) {
+	meshNs := namespace.NewOrFail(t, ctx, namespace.Config{Prefix: "mesh", Inject: true})
+	externalNs := namespace.NewOrFail(t, ctx, namespace.Config{Prefix: "external", Inject: false})
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		common.ApplyFileOrFail(ctx, meshNs.Name(), path.Join(env.IstioSrc, "samples/sleep/sleep.yaml"))
+		waitForPodsReadyOrFail(ctx, meshNs.Name(), "sleep")
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		templateParams := map[string]interface{}{
+			"codecType":  proxyHTTPVersion,
+			"tlsEnabled": proxyTLSEnabled,
+			"protocol":   serviceEntryProtocol(proxyHTTPVersion, proxyTLSEnabled),
```

Comment:
In the api spec: `HTTP version for upstream requests is determined by the service protocol defined for the proxy.`

If the SE has `protocol: HTTPS`, how is version determined?

Reply:
It's a very good question, because istio-proxy will send `http/1.1` in ALPN for `HTTP`, `HTTPS` and `TLS`. `h2` will be send only for `HTTP2` protocol - assuming that TLS origination is configured by a traffic policy. Isn't it the right approach? Is there any other way to configure `h2` for `TLS` and `HTTPS`?

Reply:
Btw. the `serviceEntryProtocol` function does not need the second argument `proxyTLSEnabled`, because TLS origination is configured in the `DestinationRule`, so I fixed it.